clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -1081,4 +1026,74 @@ boolean updateAssignmentMetadataIfNeeded(Timer timer) {
         // logic
         return updateFetchPositions(timer);
     }
+
+    @Override
+    public void subscribe(Collection<String> topics) {
+        subscribeInternal(topics, Optional.empty());
+    }
+
+    @Override
+    public void subscribe(Collection<String> topics, ConsumerRebalanceListener listener) {
+        if (listener == null)
+            throw new IllegalArgumentException("RebalanceListener cannot be null");
+
+        subscribeInternal(topics, Optional.of(listener));
+    }
+
+    @Override
+    public void subscribe(Pattern pattern) {
+        subscribeInternal(pattern, Optional.empty());
+    }
+
+    @Override
+    public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
+        if (listener == null)
+            throw new IllegalArgumentException("RebalanceListener cannot be null");
+
+        subscribeInternal(pattern, Optional.of(listener));
+    }
+
+    private void subscribeInternal(Pattern pattern, Optional<ConsumerRebalanceListener> listener) {
```

Comment:
Can we return early if the subscribed pattern is the same as the current one?

Reply:
Can we return early if the subscribed topics are the same as the existing ones?

Reply:
That's an interesting optimization. I am not sure if that will introduce subtle inconsistencies, so I'd prefer to tackle that separately.